### PR TITLE
Close the request if 100 continue failed

### PR DESCRIPTION
**Pull request by SophieGuo410**

### CODE REVIEW

File: ambry-rest/src/main/java/com/github/ambry/rest/NettyMessageProcessor.java [hunk 1]
```diff
@@ -379,9 +379,9 @@ private boolean handleContent(HttpContent httpContent) throws RestServiceExcepti
       }
       boolean isPutOrPost = request.getRestMethod().equals(RestMethod.POST) || request.getRestMethod().equals(RestMethod.PUT);
       boolean isMultipart = request.isMultipart() && requestContentFullyReceived;
-      boolean hasContinue = CONTINUE.equals(request.getArgs().get(EXPECT));
-      if (success && (!isPutOrPost || isMultipart || hasContinue)) {
-        if (hasContinue) {
+      boolean hasContinueAndIsPut = CONTINUE.equals(request.getArgs().get(EXPECT)) && request.getRestMethod().equals(RestMethod.PUT);
```
Comment (justinlin-linkedin) [inline ambry-rest/src/main/java/com/github/ambry/rest/NettyMessageProcessor.java:382]: let's create a method in RestUtils for this one.

File: ambry-rest/src/main/java/com/github/ambry/rest/NettyResponseChannel.java [hunk 2]
```diff
@@ -1111,7 +1110,12 @@ public void operationComplete(ChannelFuture future) {
           // in this case there is nothing more to write.
           if (!writeFuture.isDone()) {
             writeFuture.setSuccess();
-            completeRequest(!HttpUtil.isKeepAlive(finalResponseMetadata), false);
+            //Don't close the request when we see 100-continue in EXPECT header.
```
Comment (justinlin-linkedin) [inline ambry-rest/src/main/java/com/github/ambry/rest/NettyResponseChannel.java:1113]:
```
completeRequest(!HttpUtil.isKeepAlive(finalResponseMetadata), false, !RestUtils.isPutRequestAndExpectContinue(request));
```

Comment (SophieGuo410) [reply ambry-rest/src/main/java/com/github/ambry/rest/NettyResponseChannel.java:1113]: Sure, updated.

File: ambry-rest/src/main/java/com/github/ambry/rest/AsyncRequestResponseHandler.java [hunk 3]
```diff
@@ -299,10 +299,18 @@ public void run() {
...
+            if (requestInfo != null) {
+              logger.error("Throwable only while processing the requests" + requestInfo.restRequest, e);
+            } else {
+              logger.error("Throwable only while processing the requests", e);
+            }
+          }
           metrics.requestProcessingError.inc();
           if (requestInfo != null) {
-            onProcessingFailure(requestInfo.restRequest, requestInfo.restResponseChannel, e);
+            onProcessingFailure(requestInfo.restRequest, requestInfo.restResponseChannel, (Exception) e);
```
Comment (justinlin-linkedin) [inline ambry-rest/src/main/java/com/github/ambry/rest/AsyncRequestResponseHandler.java:313]: you can't convert this to an Exception if e is not an instance of Exception. you can probably add a line in the previous if statement to convert this throwable to exception. something like this 
```
  if (!(e instanceof Exception)) {
            metrics.requestProcessingThrowableError.inc();
            if (requestInfo != null) {
              logger.error("Throwable only while processing the requests" + requestInfo.restRequest, e);
            } else {
              logger.error("Throwable only while processing the requests", e);
            }
           e = new Exception(e);
          }
```

Comment (SophieGuo410) [reply ambry-rest/src/main/java/com/github/ambry/rest/AsyncRequestResponseHandler.java:313]: you are right, fixed it.

**Review by justinlin-linkedin — APPROVED** [review]

**Pull request merged**